dataland-backend/src/main/kotlin/org/dataland/datalandbackend/repositories/StoredCompanyRepository.kt

```diff
@@ -49,6 +51,67 @@ interface StoredCompanyRepository : JpaRepository<StoredCompanyEntity, String> {
     )
     fun searchCompanies(@Param("searchFilter") searchFilter: StoredCompanySearchFilter): List<StoredCompanyEntity>
 
+    /**
+     * A function for querying companies by search string:
+     * - searchString: If not empty, only companies that contain the search string in their name are returned
+     * (Prefix-Matches are ordered before Center-Matches,
+     * e.g. when searching for "a" Allianz will come before Deutsche Bank)
+     */
+    @Query(
+        nativeQuery = true,
+        value =
+        "WITH filtered_text_results as (" +
+            // Fuzzy-Search Company Name
+            "(SELECT company_id, company_name AS company_name," +
+            " CASE " +
+            " WHEN company_name = :#{#searchString} THEN 10" +
+            " WHEN company_name ILIKE :#{escape(#searchString)}% ESCAPE :#{escapeCharacter()} THEN 5" +
+            " ELSE 1" +
+            " END match_quality " +
+            " FROM stored_companies" +
+            " WHERE company_name ILIKE %:#{escape(#searchString)}% ESCAPE :#{escapeCharacter()}" +
+            " ORDER BY match_quality DESC, company_id LIMIT 100)" +
```

Comment:
My approach would have been to drop the ORDER BY LIMIT statements in the three WITH queries and only have a single ORDER BY LIMIT statement in the final query. I think that this is a bit faster. However, in the end you can only try it out to know and performance is the only relevant metric to decide between the two approaches. So it would be great if you could try that out and check if it would be faster.

Reply:
If we would drop the ORDER BY LIMIT statements we would get a much larger list which takes in general longer to sort as the sum of its parts. And joining three presorted lists goes super fast in linear time (if the database detects that). All I want to say is that there is a chance that this implementation is faster. But I will try it out. By "three WITH queries" you mean the portions joined via UNION, right? I only see one WITH.

Reply:
Moving the `LIMIT` into the three sub-queries significantly increases query performance (on my machine it lead to a query-time decrease from `~147ms` to ~`64ms`.

Reply:
Thanks for investigating!